### ISSUE TITLE
Watching for files being added

### ISSUE THREAD

**Issue by Globegitter** [bazelbuild/bazel-watcher#135]
When using globs, e.g. `data/**/*.txt` it would be nice if ibazel could watch this directory for new file additions and then rebuild. It would be even nicer if one could also specify to just watch for file additions, as e.g. for runfiles there is often no need to actually rebuild and the binary can just pick up the changed files due to it being a symlink.

Not sure how difficult this would be to implement, but with some thoughts/pointers on this I would be happy to look into this as well when I get some time for it.

**Comment by achew22** [comment]
If you edit a BUILD file that is part of the sources, it should trigger a requery and find files. I looked into the idea of detecting these globs and doing my own globbing. https://github.com/google/skylark/issues/109 is where I asked about it publicly and I believe the analysis in there is correct. It would be great, but unfortunately it is going to take a massive quantity of code to do it. If you have theories on ways we can achieve this, I would *LOVE* to hear your input.

Thanks!

**Comment by Globegitter** [comment]
Yeah, doing it in the way discussed in the ticket would be great i.e. that bazel watcher could parse the build files in parallel or get that information from bazel in a streaming, non-blocking fashion. But given that this is currently not possible I can only think of some kind of heuristics, where we could use buildozer to analyse some of the targets we get back from query (maybe even limiting it to filegroups), as well as the main target itself. Or maybe even simpler, if we have a target we just check if there is a filegroup of name like `<target>-globs` or `<target>-extra-watch-files`. None of these solutions are ideal but if well documented and/or opt-in then at least some people can get some file addition watching. What do you think?

Also just thinking about it in theory, what would be nice is to have something akin to the build event protocol / build event service, like loading event protocol, or analysis event protocol where bazel streams the information it finds during loading or analysis phase to a specified service with the nodes of the graph that it finds, and I suppose specifically for the watcher use-case also with some metadata about the original state of the input, i.e. this file is part of that glob.

Thinking about it further, if bazel query could give out some metadata about the original state of the rule attributes that would already help us get there and if ibazel had a flag to allow to set a different `--output_base` people who know that all of their external repositories are cached in the repository_cache could already make use of parallelization today.

I often like thinking about how could we get to a solution in the most ideal way and how can we actually get to a solution today, hence there are just a lot of thoughts spilled out and they evolved a bit as I was writing but curios to hear your thoughts on the topic.

**Comment by pward123** [comment]
In my case, I'd settle for something as simple as have ibazel kick bazel when any file is added to the workspace that is not excluded by a .gitignore.

How does ibazel currently tell bazel that something has changed?

**Comment by achew22** [comment]
> How does ibazel currently tell bazel that something has changed?

iBazel uses the [bazel query](https://docs.bazel.build/versions/master/query-how-to.html) command in [ibazel.go](https://github.com/bazelbuild/bazel-watcher/blob/master/ibazel/ibazel.go#L403) to determine the list of files that are included, transitively, in the build. All the BUILD file parsing is left to the Bazel server. The response from the Bazel server is a list of the files that are included and not the list of expressions that are used to find the files. Since we don't have access to the expressions (the input to Bazel) and only have access to the files (the output of Bazel) there is no way for us to get this data out of the current server. In order to get that data we would need to rewrite a large chunk of Bazel in Go. To quote from alandonovan's (omitted @ because I'm merely citing, not trying to drag him into this) [response to this question](https://github.com/google/skylark/issues/109#issuecomment-398200102):

> it can't be done in less than 10,000 additional lines of Go code. The difference between that functionality and a complete rewrite of 'bazel query' is really very small.

Bazel is written by a team of dozens of full time engineers. I'm a hobbyist who works on this in the fleeting hours after midnight and before bed. 10,000 lines of Go and a nearly complete rewrite is a bit out of scope for me.

That, however, doesn't mean it is out of scope for everyone. If you, or anyone reading this, want to go down the road of reimplementing the rules logic in Go, I would be extremely happy to call into a library you expose.

**Comment by jfirebaugh** [comment]
A minimal approach here might suffice. Would you accept a pull request that adds a new CLI option `-watch_for_additions=<directory>`? For each instance of that flag, ibazel would watch for new files or directories being added to the given directory (non-recursively), and treat such events as if a dependency of every target had changed.
